src/storage/src/table/state_table.rs

```diff
@@ -102,28 +111,118 @@ impl<S: StateStore> StateTable<S> {
+            self.cell_based_item = self.cell_based_streaming_iter.next().await.map_err(err)?;
+        }
+        match (self.cell_based_item.as_ref(), self.mem_table_iter.peek()) {
+            (None, None) => {
+                res = None;
+            }
+            (Some((_, row)), None) => {
+                res = Some(row.clone());
+                self.cell_based_item = self.cell_based_streaming_iter.next().await.map_err(err)?
+            }
+            (None, Some((_, row_op))) => {
+                mem_table_iter_next_flag = true;
+                match row_op {
+                    RowOp::Insert(row) => res = Some(row.clone()),
+                    RowOp::Delete(_) => {}
+                    RowOp::Update((_, new_row)) => res = Some(new_row.clone()),
+                }
+            }
+            (Some((cell_based_pk, cell_based_row)), Some((mem_table_pk, mem_table_row_op))) => {
+                let mem_table_pk_bytes =
+                    serialize_pk(mem_table_pk, &self.pk_serializer).map_err(err)?;
+                match cell_based_pk.cmp(&mem_table_pk_bytes) {
+                    Ordering::Less => {
+                        // cell_based_table_item will be return
+                        res = Some(cell_based_row.clone());
+                        self.cell_based_item =
+                            self.cell_based_streaming_iter.next().await.map_err(err)?
+                    }
+                    Ordering::Equal => {
+                        // mem_table_item will be return, while both cell_based_streaming_iter and
+                        // mem_table_iter need to execute next() once.
+                        mem_table_iter_next_flag = true;
+                        self.cell_based_item =
+                            self.cell_based_streaming_iter.next().await.map_err(err)?;
+                        match mem_table_row_op {
+                            RowOp::Insert(row) => {
+                                res = Some(row.clone());
+                            }
+                            RowOp::Delete(_) => {}
+                            RowOp::Update((_, new_row)) => res = Some(new_row.clone()),
+                        }
+                    }
+                    Ordering::Greater => {
+                        // mem_table_item will be return
+                        mem_table_iter_next_flag = true;
+                        match mem_table_row_op {
+                            RowOp::Insert(row) => res = Some(row.clone()),
+                            RowOp::Delete(_) => {}
+                            RowOp::Update((old_row, new_row)) => {
+                                debug_assert!(old_row == cell_based_row);
+                                res = Some(new_row.clone());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        if mem_table_iter_next_flag {
+            self.mem_table_iter.next();
```

Comment:
If you use a flag to determine whether to call next on mem_table, maybe use a flag on cell_based? Or vice versa if you don't use a flag?

Reply:
`mem_table_iter_next_flag ` is used here beacuse `mem_table_iter.next()` cannot be executed in "match", and perhaps  this is a limitation of the compiler.
It's a good idea to add a `cell_based_table_iter_next_flag` to unify them.